core/src/main/scala/org/apache/spark/executor/ProcfsBasedSystems.scala

```diff
@@ -0,0 +1,271 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.executor
+
+import java.io._
+import java.nio.charset.Charset
+import java.nio.file.{Files, Paths}
+
+import scala.collection.mutable
+import scala.collection.mutable.ArrayBuffer
+import scala.collection.mutable.Queue
+
+import org.apache.spark.internal.Logging
+
+
+// Some of the ideas here are taken from the ProcfsBasedProcessTree class in hadoop
+// project.
+class ProcfsBasedSystems  extends ProcessTreeMetrics with Logging {
+  val procfsDir = "/proc/"
+  var isAvailable: Boolean = isItProcfsBased
+  val pid: Int = computePid()
+  val ptree: scala.collection.mutable.Map[ Int, Set[Int]] =
+    scala.collection.mutable.Map[ Int, Set[Int]]()
+  val PROCFS_STAT_FILE = "stat"
+  var latestJVMVmemTotal: Long = 0
+  var latestJVMRSSTotal: Long = 0
+  var latestPythonVmemTotal: Long = 0
+  var latestPythonRSSTotal: Long = 0
+  var latestOtherVmemTotal: Long = 0
+  var latestOtherRSSTotal: Long = 0
+
+  createProcessTree
+
+  def isItProcfsBased: Boolean = {
+    val testing = sys.env.contains("SPARK_TESTING") || sys.props.contains("spark.testing")
+    if (testing) {
+      return true
+    }
+    try {
+      if (!Files.exists(Paths.get(procfsDir))) {
+        return false
+      }
+    }
+    catch {
+      case f: FileNotFoundException => return false
+    }
+
+    val shouldLogStageExecutorProcessTreeMetrics = org.apache.spark.SparkEnv.get.conf.
+      getBoolean("spark.eventLog.logStageExecutorProcessTreeMetrics.enabled", true)
```

Comment:
better to use the config entry `EVENT_LOG_STAGE_EXECUTOR_METRICS`

Reply:
I use this so that the user can disable process tree metrics even when the normal executor metrics are enabled. So that they have more freedom in case they feel reporting process tree metrics introduces a lot of overhead

Reply:
ok I can see having an extra config here, but then you should create a config entry for it.

Reply:
I added a config entry without the need to send config object. thanks foe the comment.